ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/views/MaterializedViewIncrementalRewritingRelVisitor.java

```diff
@@ -17,97 +17,255 @@
   }
 
-  private void check(Aggregate aggregate) {
+  public Result visit(HiveAggregate aggregate) {
+    Result result = visitChildOf(aggregate);
+    if (result.incrementalRebuildMode == NOT_AVAILABLE) {
+      return new Result(result.incrementalRebuildMode, true, -1);
+    }
+
+    Map<Integer, Set<SqlKind>> columnRefByAggregateCall = new HashMap<>(aggregate.getRowType().getFieldCount());
+
+    int countStarIndex = -1;
     for (int i = 0; i < aggregate.getAggCallList().size(); ++i) {
       AggregateCall aggregateCall = aggregate.getAggCallList().get(i);
-      if (aggregateCall.getAggregation().getKind() == SqlKind.COUNT && aggregateCall.getArgList().size() == 0) {
-        hasCountStar = true;
-        break;
+      if (aggregateCall.getAggregation().getKind() == SqlKind.COUNT &&
+          aggregateCall.getArgList().isEmpty() &&
+          !aggregateCall.isDistinct() &&
+          !aggregateCall.isApproximate()) {
+        countStarIndex = i;
+        continue;
+      }
+
+      for (Integer argIndex : aggregateCall.getArgList()) {
+        columnRefByAggregateCall.computeIfAbsent(argIndex, integer -> new HashSet<>());
+        Set<SqlKind> aggregates = columnRefByAggregateCall.get(argIndex);
+        aggregates.add(aggregateCall.getAggregation().getKind());
       }
     }
-  }
 
-  /**
-   * Starts an iteration.
-   */
-  public RelNode go(RelNode p) {
-    visit(p, 0, null);
-    return p;
-  }
+    IncrementalRebuildMode incrementalRebuildMode =
+        result.incrementalRebuildMode == INSERT_ONLY || countStarIndex == -1 ? INSERT_ONLY : AVAILABLE;
+    LOG.debug("Initial incremental rebuild mode {} input's incremental rebuild mode {} count star index {}",
+        incrementalRebuildMode, result.incrementalRebuildMode, countStarIndex);
 
-  public boolean isContainsAggregate() {
-    return containsAggregate;
-  }
+    incrementalRebuildMode = updateBasedOnAggregates(aggregate, columnRefByAggregateCall, incrementalRebuildMode);
 
-  public boolean hasAllowedOperatorsOnly() {
-    return hasAllowedOperatorsOnly;
+    return new Result(incrementalRebuildMode, true, countStarIndex);
   }
 
-  public boolean isInsertAllowedOnly() {
-    return insertAllowedOnly;
+  private IncrementalRebuildMode updateBasedOnAggregates(
+      HiveAggregate aggregate,
+      Map<Integer, Set<SqlKind>> columnRefByAggregateCall,
+      IncrementalRebuildMode incrementalRebuildMode) {
+
+    for (int i = 0; i < aggregate.getAggCallList().size(); ++i) {
+      AggregateCall aggregateCall = aggregate.getAggCallList().get(i);
+      switch (aggregateCall.getAggregation().getKind()) {
+        case COUNT:
+          if (aggregateCall.isDistinct() || aggregateCall.isApproximate()) {
+            LOG.debug("Unsupported aggregate function COUNT with distinct {} or approximate {}",
+                aggregateCall.isDistinct(), aggregateCall.isApproximate());
+            return NOT_AVAILABLE;
+          }
+
+        case SUM:
```

Comment:
nit: since these 2 aggs are not changing the supplied incrementalRebuildMode, one could remove them here.  But I suppose one advantage of keeping them here is to make it explicit that they support incremental refresh.